File: final-project/WhatsDown/app/src/main/java/com/example/whatsdown/adapter/RecentChatRecyclerAdapter.java
package com.example.whatsdown.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.whatsdown.ChatActivity;
import com.example.whatsdown.R;
import com.example.whatsdown.model.ChatMessageModel;
import com.example.whatsdown.model.ChatroomModel;
import com.example.whatsdown.model.UserModel;
import com.example.whatsdown.utils.AndroidUtil;
import com.example.whatsdown.utils.FirebaseUtil;
import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;


public class RecentChatRecyclerAdapter extends FirestoreRecyclerAdapter <ChatroomModel, RecentChatRecyclerAdapter.ChatroomModelViewHolder>{

    Context context;
    public RecentChatRecyclerAdapter(@NonNull FirestoreRecyclerOptions<ChatroomModel> options,Context context) {
        super(options);
        this.context = context;
    }

    @Override
    protected void onBindViewHolder(@NonNull ChatroomModelViewHolder holder, int position, @NonNull ChatroomModel model) {
        FirebaseUtil.getOtherUserFromChatroom(model.getUserIds())
                .get().addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        boolean lastMessageSentByMe = model.getLastMessageSenderId().equals(FirebaseUtil.currentUserId());
                        boolean isLastMessageDeleted = model.getLastMessage().isDeleted();
                        ChatMessageModel.MessageType type = model.getLastMessage().getMessageType();
                        String lastMessageText = model.getLastMessage().getMessage();

                        UserModel otherUserModel = task.getResult().toObject(UserModel.class);
                        FirebaseUtil.getOtherProfilePicStorageRef(otherUserModel.getUserId()).getDownloadUrl()
                                .addOnCompleteListener(t -> {
                                    if (t.isSuccessful()) {
                                        Uri uri = t.getResult();
                                        AndroidUtil.setProfilePic(context, uri, holder.profilePic);
                                    }
                                });
                        holder.usernameText.setText(otherUserModel.getUsername());
                        if (isLastMessageDeleted) {
                            lastMessageText = "This message was deleted.";
                        }
                        if (type != ChatMessageModel.MessageType.TEXT) {
                            lastMessageText = "📷 Attachment";
                        }

                        if (lastMessageSentByMe)
                            lastMessageText = isLastMessageDeleted ? "You deleted this message." : String.format("You: %s", lastMessageText);
                        holder.lastMessageText.setText(lastMessageText);

                        holder.lastMessageTime.setText(FirebaseUtil.timestampToString(model.getLastMessage().getTimestamp()));

                        holder.itemView.setOnClickListener(v -> {
                            //navigate to chat activity
                            Intent intent = new Intent(context, ChatActivity.class);
                            AndroidUtil.PassUserModelAsIntent(intent, otherUserModel);
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            context.startActivity(intent);
                        });
                    }
                });
    }

    @NonNull
    @Override
    public ChatroomModelViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.recent_chat_recycler_row,parent,false);
        return new ChatroomModelViewHolder(view);
    }
    class ChatroomModelViewHolder extends RecyclerView.ViewHolder{
        TextView usernameText;
        TextView lastMessageText;
        TextView lastMessageTime;
        ImageView profilePic;

        public ChatroomModelViewHolder(@NonNull View itemView) {
            super(itemView);
            usernameText = itemView.findViewById(R.id.user_name_text);
            lastMessageText = itemView.findViewById(R.id.last_message_text);
            lastMessageTime = itemView.findViewById(R.id.last_message_time_text);
            profilePic = itemView.findViewById(R.id.profile_pic_image_view);
        }
    }

}

File: Module 5/app/src/main/java/com/example/chitchat/adapter/RecentChatRecyclerAdapter.kt
package com.example.chitchat.adapter

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.chitchat.R
import com.example.chitchat.ChatActivity
import com.example.chitchat.Utils
import com.example.chitchat.models.ChatroomModel
import com.example.chitchat.models.UserModel
import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions


class RecentChatRecyclerAdapter(
    options: FirestoreRecyclerOptions<ChatroomModel>,
    var context: Context
) :
    FirestoreRecyclerAdapter<ChatroomModel, RecentChatRecyclerAdapter.ChatroomModelViewHolder>(
        options
    ) {
    protected override fun onBindViewHolder(
        holder: ChatroomModelViewHolder,
        position: Int,
        model: ChatroomModel
    ) {
        Utils.getOtherUserFromChatroom(model.userIds!!)
            .get().addOnCompleteListener { task ->
                if (task.isSuccessful()) {
                    val otherUserModel: UserModel = task.getResult().toObject(UserModel::class.java)!!
                    holder.usernameText.setText(otherUserModel.username)
                    holder.itemView.setOnClickListener { v: View? ->
                        val intent = Intent(
                            context,
                            ChatActivity::class.java
                        )
                        Utils.passUserModelAsIntent(intent, otherUserModel)
                        Utils.passChatroomModelAsIntent(intent, model)
                        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        context.startActivity(intent)
                    }
                }
            }

        if (model.userIds!!.count() > 2) {
            val stringBuilder = StringBuilder()
            Utils.getOtherUsersFromChatroom(model.userIds!!)
                .forEach { userRef ->
                    userRef.get().addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            val otherUserModel: UserModel? =
                                task.result?.toObject(UserModel::class.java)
                            if (otherUserModel != null) {
                                if (stringBuilder.isNotEmpty()) {
                                    stringBuilder.append(", ");
                                }
                                stringBuilder.append(otherUserModel.username)
                                holder.usernameText.text = stringBuilder.toString();
                            }
                        }
                    }
                }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatroomModelViewHolder {
        val view: View =
            LayoutInflater.from(context).inflate(R.layout.recent_chat_row, parent, false)
        return ChatroomModelViewHolder(view)
    }

    inner class ChatroomModelViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var usernameText: TextView

        init {
            usernameText = itemView.findViewById<TextView>(R.id.user_name_text)
        }
    }
}

